server/src/test/java/io/spine/server/delivery/given/MessageDeliveryTestEnv.java

```diff
@@ -0,0 +1,275 @@
+import io.spine.server.command.Assign;
+import io.spine.server.delivery.InProcessSharding;
+import io.spine.server.delivery.Shardable;
+import io.spine.server.delivery.Sharding;
+import io.spine.server.entity.Repository;
+import io.spine.server.transport.memory.InMemoryTransportFactory;
+import io.spine.test.aggregate.ProjectId;
+import io.spine.test.aggregate.command.AggStartProject;
+import io.spine.test.aggregate.event.AggProjectCancelled;
+import io.spine.test.aggregate.event.AggProjectPaused;
+import io.spine.test.aggregate.event.AggProjectStarted;
+import io.spine.test.aggregate.rejection.AggCannotReassignUnassignedTask;
+import io.spine.validate.StringValueVBuilder;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * An abstract base for environments, which are created to ease the message delivery testing.
+ *
+ * @author Alex Tymchenko
+ */
+public class MessageDeliveryTestEnv {
+
+    /**
+     * The time to wait until all the messages dispatched to entities
+     * are processed in several threads.
+     *
+     * <p>"Sleeping" down the main thread is a simpler choice to ensure the messages were delivered.
+     * The alternatives would imply injecting multiple mocks that would send reports
+     * down the dispatching route. Which seems to be much more complex.
+     */
+    private static final int DISPATCH_WAIT_TIME = 3000;
+
+    /** Prevents instantiation of this test environment class. */
+    private MessageDeliveryTestEnv() {
+    }
+
+    public static void setShardingTransport(InMemoryTransportFactory transport) {
+        final Sharding inProcessSharding = new InProcessSharding(transport);
+        ServerEnvironment.getInstance()
+                         .replaceSharding(inProcessSharding);
+    }
+
+    /**
+     * The time for the main thread to wait for the messages to finish dispatching in other threads.
+     *
+     * <p>The value returned should be fine for the most of current cases.
+     */
+    public static int dispatchWaitTime() {
+        return DISPATCH_WAIT_TIME;
+    }
+
+    /**
+     * The helper class used to dispatch numerous messages of a certain kind to the entities,
+     * managed by a specific repository.
+     *
+     * <p>Another feature is to check that the messages were delivered to proper shards,
+     * according to the repository sharding configuration.
+     *
+     * @param <I> the type of the entity identifiers
+     * @param <M> the type of messages to dispatch (e.g. {@link io.spine.core.Command Command}).
+     */
+    public abstract static class ParallelDispatcher<I extends Message, M extends Message> {
```

Comment:
Assuming that the test environment is big (and is not a simple bunch of data classes for tests), can we have these classes as top-level? 

It would be easier to understand.